R-package/src/lightgbm_R.cpp

```diff
@@ -25,18 +25,60 @@
 
 #define COL_MAJOR (0)
 
+#define MAX_LENGTH_ERR_MSG 1024
+char R_errmsg_buffer[MAX_LENGTH_ERR_MSG];
+struct LGBM_R_ErrorClass { SEXP cont_token; };
+void LGBM_R_save_exception_msg(const std::exception &err);
+void LGBM_R_save_exception_msg(const std::string &err);
+
 #define R_API_BEGIN() \
   try {
 #define R_API_END() } \
-  catch(std::exception& ex) { LGBM_SetLastError(ex.what()); } \
-  catch(std::string& ex) { LGBM_SetLastError(ex.c_str()); } \
-  catch(...) { LGBM_SetLastError("unknown exception"); }
+  catch(LGBM_R_ErrorClass &cont) { R_ContinueUnwind(cont.cont_token); } \
+  catch(std::exception& ex) { LGBM_R_save_exception_msg(ex); goto throw_R_errormsg; } \
+  catch(std::string& ex) { LGBM_R_save_exception_msg(ex); goto throw_R_errormsg; } \
```

Comment:
Is it possible to avoid using `goto`?

Reply:
Removed. Was added just for clarity.

Reply:
for my own knowledge @StrikerRUS , what is problematic about `goto`?

Reply:
@jameslamb I'm one of those who are against `goto` due to it leads to spaghetti code in general.
https://stackoverflow.com/a/3517765
https://en.wikipedia.org/wiki/Goto#Criticism

Reply:
ah got it, thanks!